transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
+    // So this means we are filling the 1/2 of the request.
+    // Before we add a fractional change output, this is ill-formed, we can only
+    // verify fractional output rules if there is a fractional change output,
+    // because that is how we deduce the fill fraction.
+    get_first_rules_mut(&mut tx)
+        .fractional_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(2000, 0.into()),
+        ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputsNotExpected)
+    );
+
+    // Add a fractional input, also by doubling one of the revealed tx outs.
+    // This means the fill fraction is 1/2, which we are satisfying.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(2000, 0.into()),
+    ));
+
+    // Check that the Tx is following input rules
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a limit on the change value
+    // This is larger than 1000 so we should still be valid.
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 1010;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a smaller limit that should cause things to fail
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 990;
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RealChangeOutputAmountExceedsLimit)
+    );
+
+    // Remove limit altogether, we should be valid again
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 0;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Change the frational change output to be 3000. This means the implied fill
+    // fraction is now 2/3, since the real change output only returns
+    // 1/3 of this. The tx should then be invalid because we are only filling 1/2 of
+    // the fractional output that was required and not 2/3 of it.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(3000, 0.into()),
+    ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RealOutputAmountDoesNotRespectFillFraction)
+    );
+
+    // Change the frational change output to be 500. This is less than the real
+    // change output, so that should be an error.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(500, 0.into()),
+    ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RealOutputAmountExceedsFractional)
+    );
+
+    // Change the frational change output to be 1500. This means the implied fill
```

Comment:
:thought_balloon: again it may just be me, but I think  the use of "fractional" and "change" my be misleading, throughout this pr.

All of the "fractional" values are proposed or idealized values, and the final txouts are the partial or fractional values.

Getting here I think I finally understand that txout_0 is Alice's "change" from the money she fronted

Reply:
Thanks for your thoughts, lets discuss further here: https://github.com/mobilecoinfoundation/mobilecoin/pull/2484#discussion_r979438141

The terms we use should not be a source of confusion, we should pick better terms if these names are bad.